### ISSUE TITLE
Installation error: User Namespace issue reported

### ISSUE THREAD

**Issue by StuartRus** [sandstorm-io/sandstorm#539]
On trying to install the latest sandstorm via the curl method i get the error about "user namespaces".
I ran lxc-checkconfig and every thing is ENABLED

My system:
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.2 LTS
Release:	14.04
Codename:	trusty

Any ideas how I can get it to install?

**Comment by paulproteus** [comment]
Hi Stuart! Sorry to hear you're having troubles. Can you paste the full
output of the installer, so I can get a better sense of what is going wrong?

**Comment by StuartRus** [comment]
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 53945  100 53945    0     0   7829      0  0:00:06  0:00:06 --:--:-- 14343
bash: line 332: /tmp/tmp.6Aa9BrqN4F: cannot execute binary file: Exec format error
Your kernel does not appear to be compiled with support for unprivileged user 
namespaces (CONFIG_USER_NS=y), or something else is preventing creation of user 
namespaces. This feature is critical for sandboxing. Arch Linux is known to 
ship with a kernel that disables this feature; if you are using Arch, you will 
unfortunately need to compile your own kernel (see 
https://bugs.archlinux.org/task/36969). If you are not using Arch, and don't 
know why your system wouldn't have user namespaces, please file a bug against 
Sandstorm so we can figure out what happened.
*** INSTALLATION FAILED ***
Report bugs at: http://github.com/sandstorm-io/sandstorm
stuart@linux:~$

**Comment by zarvox** [comment]
That's an `ENOEXEC` from the kernel when trying to execute the tiny userns check binary that the installer script writes out into `/tmp`.  My guess is that `/tmp` is mounted `noexec`.

@StuartRus do you mind pasting the output of `cat /proc/self/mounts | grep /tmp` to confirm that's the issue?

@paulproteus can the installer script output that file somewhere else that's less likely to be mounted noexec?

**Comment by StuartRus** [comment]
Hi
that returns nothing. 

what is line 332 in the code doing?
printf '\x7fELF\x02\x01\x01\x00kmc!!!\n\x00\x02\x00>\x00\x01\x00\x00\x00x\x00@\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@\x008\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x07\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x9d\x00\x00\x00\x00\x00\x00\x00\x9d\x00\x00\x00\x00\x00\x00\x00\x00\x10\x00\x00\x00\x00\x00\x00\xb8i\x00\x00\x00\xbf\xfe\xff\x00\x00\x0f\x05\xb8\x10\x01\x00\x00\xbf\x00\x00\x00\x10\x0f\x05H\x89\xc1\xb8<\x00\x00\x00H\x89\xcf\x0f\x05' > "$USERNS_TEST_PROGRAM"

**Comment by paulproteus** [comment]
@zarvox excellent debugging re: `ENOEXEC`.

And hi @StuartRus ,

The code on line 332 is something I wrote. What it does is write a file in `/tmp` that is a precompiled program that tests if a specific Linux kernel feature is available, and then execute it. Sorry for the opaque-seeming binary; the comments in `install.sh` explain why we do that, plus links to the source code for that. I wish I knew a better way.

Two questions:

* 1. What is the output of:

```
cat /proc/sys/kernel/unprivileged_userns_clone
```

* 2. What is the output of:

```
uname -a
```

I'm surprised you're having this problem on Ubuntu 14.04 because I believe that we whitelist Ubuntu 14.04 via the `/proc` stuff.

* 3. Are you running this within Docker or some other containerization tool? That might relate to the problems. If so, we (Sandstorm team) should figure out how to replicate that environment, so we can provide useful advice in the install script.

* 4. What is the output of:

```
unshare --user id
```

For what it's worth, if the `unshare --user id` shows this:

```
uid=65534(nobody) gid=65534(nogroup) groups=65534(nogroup)
```

then Sandstorm should work fine.

To work around this problem, I think you can edit `install.sh` -- on line 1547, delete these two lines:

```
detect_userns_clone
assert_userns_clone_works_or_can_be_made_to_work
```

Thanks for trying Sandstorm. I hope we can get to the bottom of this.

**Comment by StuartRus** [comment]
hi
stuart@linux:/usr/bin$ cat /proc/sys/kernel/unprivileged_userns_clone
cat: /proc/sys/kernel/unprivileged_userns_clone: No such file or directory

stuart@linux:/usr/bin$ uname -a
Linux linux.dyukemedia 3.13.0-55-generic #94-Ubuntu SMP Thu Jun 18 00:28:41 UTC 2015 i686 i686 i686 GNU/Linux
stuart@linux:/etc$ unshare --user id
unshare: unrecognised option '--user'

commented out the two lines and then, found out.... wait for it...

stuart@linux:~$ ./sandstorm.sh
Sorry, tha Sandstorm server currently only runs on x86_64 machines.
*** INSTALLATION FAILED ***
Report bugs at: http://github.com/sandstorm-io/sandstorm

I didnt see any requirements of needing a 64bit machine.
any work around?

**Comment by ocdtrekkie** [comment]
The reason you need 64-bit, is because all of the apps that Sandstorm runs are also going to be 64-bit binaries. So even if you were to somehow compile Sandstorm for 32-bit, your apps wouldn't work. (People have asked much the same about wanting to run it on an ARM architecture, for a Raspberry Pi or similar.)

**Comment by paulproteus** [comment]
Huge thanks to @ocdtrekkie and @StuartRus for the conversation here. I updated the install script (with @zarvox 's help) to detect the architecture first, and wrote a test for that. I also submitted a change to the https://sandstorm.io/install/ page indicating that 64-bit Linux is required.

Once those are all live, I'll mark this as closed, unless I hear a reason not to.

Sorry to you @StuartRus for the trouble you ran into. I hope you'll try Sandstorm on a 64-bit machine sometime, and that I can buy you a beverage of your choice if we ever meet up.
